# ***Credit Card Fraud Detection using Logistic Regression***

### Objective:

This machine learning project uses anonymized credit card transaction data to predict whether a transaction is fraudulent or legitimate, helping financial institutions reduce losses from fraud.

---

### Workflow Overview:

* ***Import Libraries*** – Load necessary Python packages
* ***Load & Explore Data*** – Understand the structure, class imbalance, and statistical properties
* ***Handle Class Imbalance*** – Apply under-sampling to create a balanced dataset
* ***Feature & Target Split*** – Separate input features (X) and the target label (Y)
* ***Train-Test Split*** – Split the dataset for training and evaluation
* ***Model Training*** – Train a Logistic Regression classifier
* ***Evaluate Model*** – Use accuracy, confusion matrix, and classification report
* ***Interpret Metrics*** – Gain insights from precision, recall, and F1-score

---

### Problem Type:

* Binary Classification (0 = Legitimate, 1 = Fraudulent)

---

### Dataset:

* **Source:** [Kaggle - Credit Card Fraud Detection](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud)
* **Rows:** 284,807 transactions
* **Features:**

  * `Time`: Time elapsed since the first transaction
  * `V1` to `V28`: PCA-transformed features to anonymize data
  * `Amount`: Transaction value in Euros
  * `Class`:

    * `0` → Legitimate
    * `1` → Fraudulent

---




# Step - 1
### Importing Dependencies
We'll use `NumPy`, `Pandas` for data handling, and `Scikit-learn` for model building and evaluation.

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Step - 2
### Data Collection and Preprocessing

In [3]:
# Load a CSV file directly from a ZIP archive into a DataFrame
credit_card_data = pd.read_csv('/content/creditcard.csv.zip', compression='zip')

In [4]:
# Print first few rows of data
credit_card_data.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

- ***Time***: Seconds elapsed between this transaction and the first transaction in the dataset

- ***V1 to V28***: Principal components obtained from PCA (to protect confidentiality); these are numerical and standardized

- ***Amount*** : The transaction amount (in Euros)

- ***Class***: The target variable —
  - 0 = Legitimate transaction
  - 1 = Fraudulent transaction

In [5]:
# Print the dimension of our dataset (rows,columns)
credit_card_data.shape

(284807, 31)

In [25]:
# Display basic info including data types and missing values
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [26]:
# Check for missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [27]:
# Count of legit vs fraudulent transactions
credit_card_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

***Observation***
- This dataset is highly imbalanced — legit transactions far outnumber fraudulent ones.
- Special handling (like under-sampling or over-sampling) is required before training a model.



***Seprating data for further analysis***

In [28]:
# Extract all legitimate transactions
legit = credit_card_data[credit_card_data.Class == 0 ]

In [29]:
# Extract all fraudulent transactions
fraud = credit_card_data[credit_card_data.Class == 1 ]

In [30]:
# Print the shape of both subsets
print(f"Shape of legit data:- {legit.shape}")
print(f"Shape of fraud data:- {fraud.shape}")

Shape of legit data:- (284315, 31)
Shape of fraud data:- (492, 31)


In [31]:
# Statistical summary of transaction amounts
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [19]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [32]:
# Compare mean values of features across both classes
credit_card_data.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      94838.202258  0.008258 -0.006271  ... -0.000295 -0.000131   88.291022
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

***Inference***
- The average values for several features (especially `Amount`) differ significantly between legit and fraudulent transactions.
- These differences help our model learn patterns to detect fraud effectively.

# Step - 3
### Data Analysis

***UNDER-SAMPLING:***
- Create a balanced dataset by randomly selecting 492 legitimate transactions to match the number of fraud cases


In [33]:
legit_sample = legit.sample(n = 492)    # n = number of datapoints you want to have (492 here)

In [34]:
# Combine the sampled legit and all fraud transactions into a new balanced dataset
new_dataset = pd.concat([legit_sample, fraud] , axis = 0)   # axis = 0 means df will be added row wise

In [35]:
new_dataset.head()

Time        V1        V2  ...       V28  Amount  Class
74398    55560.0  1.022768 -0.575025  ...  0.009146   56.00      0
209634  137642.0  2.018761  0.138112  ... -0.037921    1.98      0
146810   87899.0 -1.240200  1.219700  ...  0.334852    5.00      0
245269  152674.0  1.910163 -0.157871  ... -0.016769   11.40      0
75243    55967.0 -4.611536 -3.045197  ... -2.002774  557.76      0

[5 rows x 31 columns]

In [36]:
# Check if the new dataset preserves the original characteristics
new_dataset.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      95146.276423  0.139250 -0.024160  ...  0.005628 -0.007472  104.543557
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

***Note***
- The mean values in the new dataset are very similar to those in the original, which suggests the nature of our dataset is not changed.
- This particular step help to check if we got a good or bad sample; if we got a bad sample then the mean value is going to be very diffrent from previous dataframe which could mislead the model.

# Step - 4
### Feature and Target Split

In [37]:
# Features
X = new_dataset.drop(columns = 'Class' , axis = 1)

In [38]:
# Target
Y = new_dataset['Class']

In [39]:
# Preview Features
print(X)

            Time        V1        V2  ...       V27       V28  Amount
74398    55560.0  1.022768 -0.575025  ...  0.056335  0.009146   56.00
209634  137642.0  2.018761  0.138112  ... -0.063965 -0.037921    1.98
146810   87899.0 -1.240200  1.219700  ...  0.525643  0.334852    5.00
245269  152674.0  1.910163 -0.157871  ...  0.088776 -0.016769   11.40
75243    55967.0 -4.611536 -3.045197  ...  0.696922 -2.002774  557.76
...          ...       ...       ...  ...       ...       ...     ...
279863  169142.0 -1.927883  1.125653  ...  0.292680  0.147968  390.00
280143  169347.0  1.378559  1.289381  ...  0.389152  0.186637    0.76
280149  169351.0 -0.676143  1.126366  ...  0.385107  0.194361   77.89
281144  169966.0 -3.113832  0.585864  ...  0.884876 -0.253700  245.00
281674  170348.0  1.991976  0.158476  ...  0.002988 -0.015309   42.53

[984 rows x 30 columns]


In [40]:
# Preview Target
print(Y)

74398     0
209634    0
146810    0
245269    0
75243     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


# Step - 5
### Train Test Split

- Split the data into training and testing sets (80% train, 20% test)
- Using stratify=Y to maintain the class distribution in both sets

In [41]:
X_train , X_test , Y_train , Y_test = train_test_split(
    X, Y , test_size= 0.2 , random_state=2 , stratify= Y
)

In [43]:
# Display shape of the datasets
print(f"Feature data shape: {X.shape}")
print(f"X_train data shape: {X_train.shape}")
print(f"X_test data shape: {X_test.shape}")

Feature data shape: (984, 30)
X_train data shape: (787, 30)
X_test data shape: (197, 30)


# Step - 6
### Model Training

In [44]:
# Load an instance of the Logistic Regression model
model = LogisticRegression()

In [45]:
# train data on training data
model.fit(X_train, Y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

# Step - 7
### Model Evaluation

In [49]:
# Predict outcomes on the training data
train_pred = model.predict(X_train)

In [50]:
# Calculate accuracy score on training data
training_data_accuracy = accuracy_score(Y_train , train_pred)

In [51]:
# Print the training accuracy
print(f"Accuracy score of Training data:- {training_data_accuracy}")

Accuracy score of Training data:- 0.9440914866581956


In [52]:
# Predict outcomes on the test data
test_pred = model.predict(X_test)

In [53]:
# Calculate accuracy score on test data
test_data_accuracy = accuracy_score(Y_test , test_pred)

In [54]:
# Print the test accuracy
print(f"Accuracy score of Test data:- {test_data_accuracy}")

Accuracy score of Test data:- 0.9441624365482234


In [56]:
from sklearn.metrics import confusion_matrix, classification_report

# Predict on test data
test_pred = model.predict(X_test)

# Confusion Matrix
conf_matrix = confusion_matrix(Y_test, test_pred)
print("Confusion Matrix:")
print(conf_matrix)

# Classification Report
print("\nClassification Report:")
print(classification_report(Y_test, test_pred, target_names=["Legit", "Fraud"]))

Confusion Matrix:
[[95  4]
 [ 7 91]]

Classification Report:
              precision    recall  f1-score   support

       Legit       0.93      0.96      0.95        99
       Fraud       0.96      0.93      0.94        98

    accuracy                           0.94       197
   macro avg       0.94      0.94      0.94       197
weighted avg       0.94      0.94      0.94       197



**Evaluation Summary**

- **Confusion Matrix**

```
               Predicted
             | Legit | Fraud
  -----------|-------|-------
  Actual Legit |  95  |   4
  Actual Fraud |   7  |  91
```

* **True Positives (TP)**: 91 → Correctly predicted fraud
  * **True Negatives (TN)**: 95 → Correctly predicted legit
  * **False Positives (FP)**: 4 → Legit predicted as fraud
  * **False Negatives (FN)**: 7 → Fraud predicted as legit

---

- **Classification Report Insights**

| Metric    | Legit | Fraud |
| --------- | ----- | ----- |
| Precision | 0.93  | 0.96  |
| Recall    | 0.96  | 0.93  |
| F1-score  | 0.95  | 0.94  |

#### Accuracy: `94%`

* The model correctly predicted 94% of all transactions.

#### Precision (Fraud): `0.96`

* Out of all transactions predicted as fraud, 96% were actually fraudulent.

#### Recall (Fraud): `0.93`

* Out of all actual frauds, the model identified 93%.

#### F1-score:

* A balanced score between precision and recall — very strong for both classes.

